coreneuron/utils/progressbar/progressbar.c

```diff
@@ -118,35 +108,36 @@ void progressbar_free(progressbar* bar) {
 void progressbar_update(progressbar* bar, unsigned long value, double t) {
     bar->value = value;
     bar->t = t;
-    // Progress bar will be drawn every BAR_DRAW_INTERVAL seconds
-    // for the first BAR_DRAW_INIT_PERIOD seconds (Initial period) and
-    // then choose between drawing every BAR_DRAW_INTERVAL seconds or
-    // draw the bar BAR_DRAW_REMAINING_UPDATES times, depending on
-    // which option draws the progress bar less times.
+    // Sample ETA value throughout the simulation to calculate the time intervals
+    // between progress bar redraws based on the max number of progress
+    // bar drawings wanted. ETA sampling makes the time intervals for redrawing adapt to the time
+    // the simulation takes while avoiding ETA extreme values in the first simulation steps or the
+    // end of the simulation. Up until the first ETA sample, progress bar is redrawn every 2
+    // seconds. The total number of progress bars drawn may surpass the BAR_DRAW_COUNT_MAX, but only
+    // by a little due to the ETA forecasting variability.
     time_t cur_time = time(NULL);
     int sim_time = difftime(cur_time, bar->start);
-    int draw_init_period = sim_time <= BAR_DRAW_INIT_PERIOD;
     int eta_s = progressbar_remaining_seconds(bar);
 
-    if (draw_init_period) {
-        bar->eta_init_period_last = eta_s;
-        bar->value_init_period_last = bar->value;
-    }
+    int sample_interval = bar->max / NUMBER_OF_ETA_SAMPLES;
+
+    int is_more_than_sample_interval = (bar->value - bar->prev_sample_value) >= sample_interval;
 
-    int draw_remaining_period_with_freq =
-        bar->eta_init_period_last / BAR_DRAW_INTERVAL <= BAR_DRAW_COUNT_REMAINING;
-    int is_more_than_time_interval = difftime(cur_time, bar->prev_t) >= BAR_DRAW_INTERVAL;
-    int is_more_than_steps_interval =
-        (bar->value - bar->prev_value) >=
-        (bar->max - bar->value_init_period_last) / BAR_DRAW_COUNT_REMAINING;
+    if (is_more_than_sample_interval) {
+        bar->prev_sample_value = bar->value;
+        bar->draw_time_interval = (eta_s) / (BAR_DRAW_COUNT_MAX - bar->drawn_count);
+        if (bar->draw_time_interval < 1)
+            bar->draw_time_interval = 1;  // avoid output in every time step
```

Comment:
Better to always use `{ }` even in single line statements. (easy to add a second line and screw the logic :p )

Reply:
Thank you very much for the extensive and detailed comments @ferdonline . I really appreciate them, so that I can learn as much as I can. 
Just some further development on your thoughts:
> * 1000 samples seems a lot for me. Maybe ask one of our mathematicians, but I have the impression that to get a good average from a distribution we need actually very few samples (kind, less than 50...).

It's true that a large number of samples is not needed to get a good understanding of the ETA. However, I kept the sampling number high, so that in experiments with a large number of timesteps the first sampling is done quickly to avoid a large number of bars being redrawn(every 2 seconds) during this time. 
> * If the simulation gets heavier and heavier with time, even a lot of samples won't represent the mean. (Maybe Pramod knows if that's a common scenario or not). In case it is, a moving average would be better (of course, this is if we really want to be fancy ;) )
> * If we can stay on the simple side, maybe we could have `sample_interval = draw_time_interval` and simplify those if's.
You might be right in that. I want to get some more feedback on the different kinds of simulation behaviors to check if the sampling is okay or there are any problems with than in any case.

Thank you again and I will also correct the single line if-statement brackets. Good observation :+1:

Reply:
another advantage of `sample_interval = draw_time_interval` is that we compute less ETA samples for small simulations.
For the "running average" consider simply:
- `projected_ETA = ETA_according_to_last_interval * 0.5 + previous_projected_ETA * 0.5`

If it works well it's what I call elegant: little code, great value! :)
What do you think?

Reply:
Thanks for your input again Fernando.
`sample_interval = draw_time_interval` is a very good choice indeed I think and I will add it after testing it.
Regarding the moving average, after some consideration I have found it may not be that necessary but please correct me if I am wrong as I may not have understood correctly the idea. ETA is the `simulation_time / number_of_timesteps_passed * number_of_time_steps_remaining`, so it is already based on the average time a simulation step takes and every time it is calculated it is already making a "better" forecast, as it is based in more timesteps. According to my understanding the moving average would have a greater impact in the resulting ETA if there was a huge difference in the average time needed to calculate a simulation timestep before and after the previous ETA calculation.  If I had understood correctly the way you have written your formula, I think that it should be `projected_ETA = (ETA_according_to_last_interval / current_simulation_step * 0.5 + previously_projected_ETA / previous_ETA_calc_simulation_step * 0.5) * simulation_steps_remaining` as ETA is time needed to finish the simulation, which is something different in every timestep.
Sorry for the long text and I hope I made my thoughts clear. Please tell me if I made any mistake in my train of thought or you want to develop further any point, as I see that the progress bar drawing has captivated the interest of both of us :P

Reply:
Very clear, thanks! I see you introduced the elapsed time component in the weighted average, which indeed was required to be correct! But maybe with `ETA_according_to_last_interval` you meant the "time of the last sample", no? But yeah, that's about it. 
If it's a matter of changing the formula and you don't need to store more vars I'd give it a shot, otherwise keep the simple implementation.

Reply:
@ferdonline I would like to thank you very much for all your feedback! :+1: 
After discussing about the moving average with Pramod we reached a conclusion that the moving average is too much, as the time needed to calculate a single timestep is not changing a lot throughout the simulation, so it is not needed to be implemented for now.